src/Orleans.Core.Abstractions/Runtime/RequestContext.cs

```diff
@@ -20,34 +20,49 @@ namespace Orleans.Runtime
     /// </remarks>
     public static class RequestContext
     {
-        /// <summary>
-        /// Gets or sets a value indicating whether <c>Trace.CorrelationManager.ActivityId</c> settings should be propagated into grain calls.
-        /// </summary>
-        public static bool PropagateActivityId { get; set; }
-
-        internal const string CALL_CHAIN_REQUEST_CONTEXT_HEADER = "#RC_CCH";
-        internal const string E2_E_TRACING_ACTIVITY_ID_HEADER = "#RC_AI";
+        internal const string CALL_CHAIN_REENTRANCY_HEADER = "#CCR";
         internal const string PING_APPLICATION_HEADER = "Ping";
 
         internal static readonly AsyncLocal<ContextProperties> CallContextData = new AsyncLocal<ContextProperties>();
 
-        /// <summary>Gets or sets an activity ID that can be used for correlation.</summary>
-        public static Guid ActivityId
+        public static Guid ReentrancyId
         {
-            get { return (Guid)(Get(E2_E_TRACING_ACTIVITY_ID_HEADER) ?? Guid.Empty); }
+            get => Get(CALL_CHAIN_REENTRANCY_HEADER) is Guid guid ? guid : Guid.Empty;
             set
             {
                 if (value == Guid.Empty)
                 {
-                    Remove(E2_E_TRACING_ACTIVITY_ID_HEADER);
+                    Remove(CALL_CHAIN_REENTRANCY_HEADER);
                 }
                 else
-                { 
-                    Set(E2_E_TRACING_ACTIVITY_ID_HEADER, value);
+                {
+                    Set(CALL_CHAIN_REENTRANCY_HEADER, value);
                 }
             }
         }
 
+        public static ConfiguredCallChain AllowCallChainReentrancy()
```

Comment:
It feels a bit weird to me that it will return with an empty Guid if the call chain wasn't set before. For SuppressCallChainReentrancy it makes more sense though